dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -2246,50 +2246,79 @@ bool RtpsUdpDataLink::separate_message(EntityId_t entity)
 namespace {
 
 struct BundleHelper {
+  const size_t initial_size =
+#ifdef OPENDDS_SECURITY
+    RtpsUdpSendStrategy::MaxSecureFullMessageLeadingSize;
+#else
+    0
+#endif
+    ;
+
+  const size_t max_size_handicap =
+#ifdef OPENDDS_SECURITY
+    RtpsUdpSendStrategy::MaxSecureFullMessageFollowingSize
+#else
+    0
+#endif
+    ;
+
   BundleHelper(size_t max_bundle_size, OPENDDS_VECTOR(size_t)& meta_submessage_bundle_sizes)
-  : max_bundle_size_(max_bundle_size)
-  , size_(0)
-  , padding_(0)
-  , prev_size_(0)
-  , prev_padding_(0)
+  : max_bundle_size_(max_bundle_size - max_size_handicap)
+  , size_(initial_size)
   , meta_submessage_bundle_sizes_(meta_submessage_bundle_sizes)
+  , first_add_to_bundle_(true)
   {
   }
 
   void end_bundle() {
-    meta_submessage_bundle_sizes_.push_back(size_ + padding_);
-    size_ = 0; padding_ = 0; prev_size_ = 0; prev_padding_ = 0;
+    meta_submessage_bundle_sizes_.push_back(size_);
+    size_ = initial_size;
   }
 
   template <typename T>
-  void push_to_next_bundle(const T&) {
-    meta_submessage_bundle_sizes_.push_back(prev_size_ + prev_padding_);
-    size_ -= prev_size_; padding_ -= prev_padding_; prev_size_ = 0; prev_padding_ = 0;
-  }
+  bool add_to_bundle(const T& val, size_t& submessage_length) {
+    submessage_length = 0;
+    ACE_DEBUG((LM_DEBUG, "(%P|%t) add_to_bundle: %C\n", gen_OpenDDS_RTPS_SubmessageKind_names[val.smHeader.submessageId]));
 
-  template <typename T>
-  bool add_to_bundle(const T& val) {
-    prev_size_ = size_;
-    prev_padding_ = padding_;
-    gen_find_size(val, size_, padding_);
-    if ((size_ + padding_) > max_bundle_size_) {
-      push_to_next_bundle(val);
-      return false;
+    if (first_add_to_bundle_) {
```

Comment:
What happens differently based on `first_add_to_bundle_`?

Reply:
It's just for logging that it is the first add in the series of bundles so it's easy to see in the logs. This will be removed along with the extra debug messages.